### PR TITLE
Pg stat statements reset for neon superuser

### PR DESCRIPTION
**Pull request by Bodobolero**
## Problem

Extension pg_stat_statements has function pg_stat_statements_reset().
In vanilla Postgres this function can only be called by superuser role or other users/roles explicitly granted.
In Neon no end user can use superuser role.
Instead we have neon_superuser role.
We need to grant execute on pg_stat_statements_reset() to neon_superuser

## Summary of changes

Modify the Postgres v14, v15, v16 contrib in our compute docker file to grant execute on pg_stat_statements_reset() to neon_superuser. (Modifying it in our docker file is preferable to changes in neondatabase/postgres because we want to limit the changes in our fork that we have to carry with each new version of Postgres).

Note that the interface of proc/function pg_stat_statements_reset changed in pg_stat_statements version 1.7

So for versions up to and including 1.6 we must

`GRANT EXECUTE ON FUNCTION pg_stat_statements_reset() TO neon_superuser;`

and for versions starting from 1.7 we must

`GRANT EXECUTE ON FUNCTION pg_stat_statements_reset(Oid, Oid, bigint) TO neon_superuser;`

If we just use `GRANT EXECUTE ON FUNCTION pg_stat_statements_reset() TO neon_superuser;` for all version this results in the following error for versions 1.7+:

```sql
neondb=> create extension pg_stat_statements;
ERROR:  function pg_stat_statements_reset() does not exist
```

## Checklist before requesting a review

- [x ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [x ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

## I have run the following test and could now invoke pg_stat_statements_reset() using default user

```bash
(neon) peterbendel@Peters-MBP neon % kubectl get pods | grep compute-quiet-mud-88416983       
compute-quiet-mud-88416983-74f4bf67db-crl4c            3/3     Running     0          7m26s
(neon) peterbendel@Peters-MBP neon % kubectl set image deploy/compute-quiet-mud-88416983 compute-node=neondatabase/compute-node-v15:7307610371
deployment.apps/compute-quiet-mud-88416983 image updated
(neon) peterbendel@Peters-MBP neon % psql postgresql://peterbendel:<secret>@ep-bitter-sunset-73589702.us-east-2.aws.neon.build/neondb     
psql (16.1, server 15.5)
SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, compression: off)
Type "help" for help.

neondb=> select version();
                                              version                                              
---------------------------------------------------------------------------------------------------
 PostgreSQL 15.5 on x86_64-pc-linux-gnu, compiled by gcc (Debian 10.2.1-6) 10.2.1 20210110, 64-bit
(1 row)

neondb=> create extension pg_stat_statements;
CREATE EXTENSION

neondb=> select pg_stat_statements_reset();
 pg_stat_statements_reset 
--------------------------
 
(1 row)
```

### CODE REVIEW

**Comment by bayandin** [comment]
You don't need to use `approved-for-ci-run` label, we use it for running CI for external contributors (because we don't want to run CI automatically for untrusted code)

See https://github.com/neondatabase/neon/blob/main/CONTRIBUTING.md#how-to-run-a-ci-pipeline-on-pull-requests-from-external-contributors

File: vendor/postgres-v15
Comment (bayandin) [inline vendor/postgres-v15:1]: > Submodule [postgres-v15](https://github.com/neondatabase/postgres) updated [245 files](https://github.com/neondatabase/postgres/compare/24333abb81a9ecae4541019478f0bf7d0b289df7...d9d6957148ec46b0cadae49e2c3a9e5f7573e4b5)

This doesn't look right

File: vendor/postgres-v14
Comment (bayandin) [inline vendor/postgres-v14:1]: This contains a couple of commits from @save-buffer, don't know if we want to add it. But if we want, probably it should be in a separate PR

**Comment by bayandin** [comment]
https://github.com/neondatabase/neon/actions/runs/7302179882/job/19900348069?pr=6232 job failed: 
```
Expected postgres-v14 rev to be at '0bb356aa0cd1582112926fbcf0b5370222c2db6d', but it is at '7de983ea08e2d9c11be79093b9620adfc9b479e2'
Expected postgres-v15 rev to be at '24333abb81a9ecae4541019478f0bf7d0b289df7', but it is at 'd9d6957148ec46b0cadae49e2c3a9e5f7573e4b5'
Expected postgres-v16 rev to be at '863b71572bc441581efb3bbee2ad18af037be1bb', but it is at '6700d915d6f9e2c762f51d6dd0744807902c0ce5'
Please update vendors/revisions.json if these changes are intentional
```

File: vendor/postgres-v14
Comment (vadim2404) [reply vendor/postgres-v14:1]: yes, we want them, but it should be a separate PR

File: compute_tools/src/spec.rs [hunk 1]
```diff
@@ -678,6 +678,13 @@ pub fn handle_extensions(spec: &ComputeSpec, client: &mut Client) -> Result<()>
             let query = "CREATE EXTENSION IF NOT EXISTS pg_stat_statements";
             info!("creating system extensions with query: {}", query);
             client.simple_query(query)?;
+            // since customers can not use superuser we grant execute to neon_superuser
+            let query = "GRANT EXECUTE ON FUNCTION pg_stat_statements_reset() TO neon_superuser";
```
Comment (vadim2404) [inline compute_tools/src/spec.rs:682]: this code will be executed only for users who have this feature flag. we didn’t expose it for the users (you might see the Insights tab in the console, this part of code is used for this particular feature). 

It means that your changes will not help for general use case

**Review by vadim2404 — CHANGES_REQUESTED** [review]

File: compute_tools/src/spec.rs (hunk 1 above)
Comment (Bodobolero) [reply compute_tools/src/spec.rs:682]: I marked this PR as draft and did not yet assign a reviewer, this is WIP

Comment (Bodobolero) [reply compute_tools/src/spec.rs]: New solution doesn't modify spec.rs

File: Dockerfile.compute-node [hunk 2]
```diff
@@ -48,7 +48,20 @@ RUN cd postgres && \
     echo 'trusted = true' >> /usr/local/pgsql/share/extension/pgrowlocks.control && \
     echo 'trusted = true' >> /usr/local/pgsql/share/extension/pgstattuple.control && \
     echo 'trusted = true' >> /usr/local/pgsql/share/extension/refint.control && \
-    echo 'trusted = true' >> /usr/local/pgsql/share/extension/xml2.control
+    echo 'trusted = true' >> /usr/local/pgsql/share/extension/xml2.control && \
+    old_list="pg_stat_statements--1.0--1.1.sql pg_stat_statements--1.1--1.2.sql pg_stat_statements--1.2--1.3.sql pg_stat_statements--1.3--1.4.sql pg_stat_statements--1.4--1.5.sql pg_stat_statements--1.4.sql pg_stat_statements--1.5--1.6.sql"; \
```
Comment (vadim2404) [inline Dockerfile.compute-node:52]: you can get the list of this files

```bash
find . -name "*.sql" -type f
```

Comment (vadim2404) [reply Dockerfile.compute-node:52]: oh, I got what did you want to achieve

Comment (Bodobolero) [reply Dockerfile.compute-node:52]: yes, the solution explicitly lists files up to and including version 1.6 of pg_stat_statements and is generic for all future versions (not in old_list).

Comment (vadim2404) [reply Dockerfile.compute-node:52]: I’d recommend to leave the comment in source code. Because, I assume, that someone can try to refactor these lines as I wanted to

File: Dockerfile.compute-node [hunk 3]
```diff
@@ -48,7 +48,20 @@ RUN cd postgres && \
...
+        if echo "$old_list" | grep -q -F "$filename"; then \
+            echo 'GRANT EXECUTE ON FUNCTION pg_stat_statements_reset() TO neon_superuser;' >> $file; \
+        fi; \
+    done; \
+    for file in /usr/local/pgsql/share/extension/pg_stat_statements--*.sql; do \
+        filename=$(basename "$file"); \
+        if ! echo "$old_list" | grep -q -F "$filename"; then \
+            echo 'GRANT EXECUTE ON FUNCTION pg_stat_statements_reset(Oid, Oid, bigint) TO neon_superuser;' >> $file; \
+        fi; \
+    done
```
Comment (bayandin) [inline Dockerfile.compute-node:64]: - Won't it be enough, to add it only to one of the files (to 1.7, for example)? If it's not enough, then maybe we can put it into a single loop (I haven't checked it)

```suggestion
    for file in /usr/local/pgsql/share/extension/pg_stat_statements--*.sql; do \
        filename=$(basename "$file"); \
        if echo "$old_list" | grep -q -F "$filename"; then \
            echo 'GRANT EXECUTE ON FUNCTION pg_stat_statements_reset() TO neon_superuser;' >> $file; \
        else \
            echo 'GRANT EXECUTE ON FUNCTION pg_stat_statements_reset(Oid, Oid, bigint) TO neon_superuser;' >> $file; \
        fi; \
    done; \
```

- Also, let's add a comment about why do we need to do it here

**Review by bayandin — COMMENTED** [review]
CI status is not displayed for some reason, nevertheless, I can see it in the list of workflow runs: https://github.com/neondatabase/neon/actions/workflows/build_and_test.yml?query=branch%3Apg_stat_statements_reset_for_neon_superuser

Let's try to rebase the branch to see if the fresh runs are displayed correctly

File: Dockerfile.compute-node (hunk 3 above)
Comment (Bodobolero) [reply Dockerfile.compute-node:64]: It is not enough to add it to a single version because you can create an extension with a specific version, for example

`create extension pg_stat_statements with version '1.4';`

You can also upgrade to a specific version using

`ALTER EXTENSION pg_stat_statements UPDATE TO '1.6';`

so customers can run the extension in their preferred version, also if they pg_dump / pg_restore they may create the extension in a given version.

So we need to grant execute to neon_superuser for all possible versions so that each version can be used with neon.

I wil add a comment why we do it here

**Review by bayandin — APPROVED** [review]
